File: windows/tests/Elevate.App.Tests/AppModelPropagationTests.cs
using Elevate.App.Tests.Support;
using Elevate.App.ViewModels;
using Elevate.Core.Coordination;
using Elevate.Core.Models;
using Elevate.Core.Storage;
using Elevate.Core.Tests.Support;
using FluentAssertions;

namespace Elevate.App.Tests;

/// <summary>
/// The panel's side of #181: a row is not "ready" because PIM says active. Port of the macOS
/// <c>AppModelPropagationTests</c>.
/// </summary>
public class AppModelPropagationTests
{
    private const string GlobalAdmin = "62e90394-69f5-4237-9190-012177145e10";

    private static RoleKey EntraKey => Sample.Key(new EntraDirectoryScope(GlobalAdmin, "/"));

    private static AppState State() => new() { Identities = [Sample.Identity()], Tenants = [Sample.Tenant()] };

    private static ActivationOutcome Activated(RoleKey key, DateTimeOffset? started = null) =>
        new(key, new ActivationResult.Activated(new ActiveAssignment(
            key, "a1", started ?? DateTimeOffset.UtcNow, DateTimeOffset.UtcNow.AddHours(1), AssignmentStatus.Active)));

    /// <summary>
    /// A model whose probe answers from <paramref name="token"/>'s claims. The pacing is a
    /// millisecond so a test that waits for an answer gets one at once; pass
    /// <paramref name="probeAfter"/> to hold the probe off instead, for a test about the state
    /// before any probe has run.
    /// </summary>
    private static async Task<TestModel> ModelAsync(
        string? token, RecordingNotifier? notifier = null, TimeSpan? probeAfter = null)
    {
        var tokens = new FakeTokenProvider { RefreshedToken = token };
        var test = await TestModel.BootstrappedAsync(State(), online: true, tokens: tokens, notifier: notifier);
        test.Model.PropagationFirstInterval = probeAfter ?? TimeSpan.FromMilliseconds(1);
        test.Model.PropagationMaxInterval = probeAfter ?? TimeSpan.FromMilliseconds(1);
        return test;
    }

    /// <summary>
    /// Waits for <paramref name="condition"/> rather than for a length of time. The model reports a
    /// probe from a background task, so anything it triggers — a row changing, a notification —
    /// lands a moment after the call that started it returns.
    /// </summary>
    private static async Task EventuallyAsync(Func<bool> condition, string what)
    {
        var deadline = DateTimeOffset.UtcNow.AddSeconds(5);
        while (!condition())
        {
            if (DateTimeOffset.UtcNow > deadline)
            {
                throw new TimeoutException($"Timed out waiting for {what}.");
            }

            await Task.Delay(5);
        }
    }

    /// <summary>Waits for the watch on <paramref name="key"/> to settle, rather than for a fixed time.</summary>
    private static async Task SettledAsync(AppModel model, RoleKey key)
    {
        var deadline = DateTimeOffset.UtcNow.AddSeconds(5);
        while (model.Propagation.TryGetValue(key, out var state) && state == PropagationState.Propagating)
        {
            if (DateTimeOffset.UtcNow > deadline)
            {
                throw new TimeoutException($"The probe on {key.Scope.Kind} never settled.");
            }

            await Task.Delay(5);
        }
    }

    [Fact]
    public async Task ARowIsPropagatingTheMomentItIsActivated()
    {
        // The probe is held off, so this is the state WatchPropagation sets before asking anything.
        using var test = await ModelAsync(token: null, probeAfter: TimeSpan.FromMinutes(5));
        var key = EntraKey;
        test.Model.Active[key] = new ActiveAssignment(key, "a1", DateTimeOffset.UtcNow, null, AssignmentStatus.Active);

        test.Model.WatchPropagation([Activated(key)]);

        test.Model.Propagation[key].Should().Be(PropagationState.Propagating);
        test.Model.PropagationNote(key).Should().Be("propagating (~3 min)");
        test.Model.PropagationTooltip(key).Should().Contain("checking");
    }

    [Fact]
    public async Task AConfirmedProbeClearsTheRowAndNotifies()
    {
        var notifier = new RecordingNotifier();
        using var test = await ModelAsync(Jwt.WithRoles(GlobalAdmin), notifier);
        var key = EntraKey;
        // Started well before now, so it is not the "nobody waited for this" case.
        var started = DateTimeOffset.UtcNow.AddMinutes(-1);
        test.Model.Active[key] = new ActiveAssignment(key, "a1", started, null, AssignmentStatus.Active);

        test.Model.WatchPropagation([Activated(key, started)]);
        await SettledAsync(test.Model, key);

        test.Model.Propagation.Should().NotContainKey(key);
        test.Model.PropagationNote(key).Should().BeNull();
        // The notification is raised without being awaited, so it lands just after the state does.
        await EventuallyAsync(() => notifier.Posted.Count > 0, "the ready notification");
        notifier.Posted.Should().ContainSingle().Which.Title.Should().Contain("is ready");
    }

    [Fact]
    public async Task ARoleReadyStraightAwayIsNotWorthANotification()
    {
        var notifier = new RecordingNotifier();
        using var test = await ModelAsync(Jwt.WithRoles(GlobalAdmin), notifier);
        var key = EntraKey;
        test.Model.Active[key] = new ActiveAssignment(key, "a1", DateTimeOffset.UtcNow, null, AssignmentStatus.Active);

        test.Model.WatchPropagation([Activated(key)]);
        await SettledAsync(test.Model, key);
        // Nothing to wait for here, so give the notification every chance to appear and prove it does not.
        await Task.Delay(100);

        test.Model.Propagation.Should().NotContainKey(key);
        notifier.Posted.Should().BeEmpty();
    }

    [Fact]
    public async Task AnUnobservableRoleGoesBackToAPlainActiveRow()
    {
        // A role scoped to an administrative unit is not carried in a token, so there is nothing to say.
        using var test = await ModelAsync(Jwt.WithRoles());
        var key = Sample.Key(new EntraDirectoryScope(GlobalAdmin, "/administrativeUnits/au1"));
        test.Model.Active[key] = new ActiveAssignment(key, "a1", DateTimeOffset.UtcNow, null, AssignmentStatus.Active);

        test.Model.WatchPropagation([Activated(key)]);
        await SettledAsync(test.Model, key);

        test.Model.Propagation.Should().NotContainKey(key);
        test.Model.PropagationNote(key).Should().BeNull();
    }

    [Fact]
    public async Task DeactivatingWhileItPropagatesDropsTheWatch()
    {
        using var test = await ModelAsync(token: null, probeAfter: TimeSpan.FromMinutes(5));
        var key = EntraKey;
        test.Model.Active[key] = new ActiveAssignment(key, "a1", DateTimeOffset.UtcNow, null, AssignmentStatus.Active);
        test.Model.WatchPropagation([Activated(key)]);

        test.Model.StopWatchingPropagation(key);

        test.Model.Propagation.Should().NotContainKey(key);
    }

    [Fact]
    public async Task OnlyActivatedOutcomesAreWatched()
    {
        using var test = await ModelAsync(token: null);
        var key = EntraKey;
        var pending = new ActiveAssignment(key, "a1", DateTimeOffset.UtcNow, null, AssignmentStatus.PendingApproval);

        test.Model.WatchPropagation([new ActivationOutcome(key, new ActivationResult.PendingApproval(pending))]);

        test.Model.Propagation.Should().BeEmpty();
    }

    [Fact]
    public async Task ARoleThatIsNotBeingWatchedIsNotPropagating()
    {
        // The zero value of PropagationState is Propagating, so a lookup that falls back to the
        // default would mark every plain active row as propagating.
        using var test = await ModelAsync(token: null);

        test.Model.Propagation.Should().BeEmpty();
        test.Model.PropagationNote(EntraKey).Should().BeNull();
        test.Model.PropagationTooltip(EntraKey).Should().BeNull();
    }

    [Fact]
    public async Task TheUnconfirmedNoteNamesWhatToDoAboutIt()
    {
        using var test = await ModelAsync(token: null);
        var key = EntraKey;
        test.Model.Propagation[key] = PropagationState.Unconfirmed;

        test.Model.PropagationNote(key).Should().Be("not in effect yet");
        test.Model.PropagationTooltip(key).Should().Be(PropagationHints.LikelyCause(RoleScopeKind.EntraDirectory));
    }

}

File: macos/Tests/ElevateAppTests/AppModelPropagationTests.swift
import Foundation
import Testing
import ElevateCore
@testable import Elevate

/// The panel's side of #181: a row is not "ready" because PIM says active.
@MainActor
struct AppModelPropagationTests {
    private static let globalAdmin = "62e90394-69f5-4237-9190-012177145e10"
    private static var entraKey: RoleKey {
        Sample.key(.entraDirectory(roleDefinitionId: globalAdmin, directoryScopeId: "/"))
    }

    /// A model whose probe answers from `token`'s claims, paced so a test does not sit through the real interval.
    private func makePropagationModel(token: String?, notifier: any ExpiryNotifying = NoopNotifier()) async -> AppModel {
        var state = AppState()
        state.identities = [Sample.identity()]
        state.tenants = [Sample.tenant()]
        let tokens = FakeTokenProvider()
        await tokens.setRefreshedToken(token)
        let model = await makeModel(state: state, online: true, tokens: tokens, notifier: notifier)
        model.propagationFirstInterval = 0.001
        model.propagationMaxInterval = 0.001
        return model
    }

    private func activated(_ key: RoleKey, started: Date = .now) -> ActivationOutcome {
        ActivationOutcome(roleKey: key, result: .activated(ActiveAssignment(
            roleKey: key, assignmentId: "a1", startDateTime: started,
            endDateTime: Date().addingTimeInterval(3600), status: .active)))
    }

    /// Waits for the watch on `key` to settle, rather than for a fixed time.
    private func settled(_ model: AppModel, _ key: RoleKey) async throws {
        for _ in 0..<200 {
            if model.propagation[key] != .propagating { return }
            try await Task.sleep(for: .milliseconds(25))
        }
        Issue.record("The probe on \(key.scope.kind) never settled.")
    }

    @Test func aRowIsPropagatingTheMomentItIsActivated() async throws {
        // No token to read, so the probe cannot confirm — but the row must say so before it asks.
        let model = await makePropagationModel(token: nil)
        defer { cleanup(model) }
        let key = Self.entraKey
        model.active[key] = Sample.assignment(key)

        model.watchPropagation([activated(key)])

        #expect(model.propagation[key] == .propagating)
        #expect(model.propagationNote(for: key) == "propagating (~3 min)")
        #expect(model.propagationTooltip(for: key)?.contains("checking") == true)
        #expect(model.isPropagating(key))
    }

    @Test func aConfirmedProbeClearsTheRowAndNotifies() async throws {
        let notifier = RecordingNotifier()
        let model = await makePropagationModel(token: Jwt.withRoles(Self.globalAdmin), notifier: notifier)
        defer { cleanup(model) }
        let key = Self.entraKey
        // Started well before now, so it is not the "nobody waited for this" case.
        let started = Date().addingTimeInterval(-60)
        model.active[key] = ActiveAssignment(roleKey: key, assignmentId: "a1", startDateTime: started,
                                             endDateTime: nil, status: .active)

        model.watchPropagation([activated(key, started: started)])
        try await settled(model, key)

        #expect(model.propagation[key] == nil)
        #expect(model.propagationNote(for: key) == nil)
        // The notification is posted from a detached task; give it a turn.
        try await Task.sleep(for: .milliseconds(50))
        #expect(await notifier.notifications.first?.title.contains("is ready") == true)
    }

    @Test func aRoleReadyStraightAwayIsNotWorthANotification() async throws {
        let notifier = RecordingNotifier()
        let model = await makePropagationModel(token: Jwt.withRoles(Self.globalAdmin), notifier: notifier)
        defer { cleanup(model) }
        let key = Self.entraKey
        model.active[key] = Sample.assignment(key)

        model.watchPropagation([activated(key)])
        try await settled(model, key)
        try await Task.sleep(for: .milliseconds(50))

        #expect(model.propagation[key] == nil)
        #expect(await notifier.notifications.isEmpty)
    }

    @Test func anUnobservableRoleGoesBackToAPlainActiveRow() async throws {
        // A role scoped to an administrative unit is not carried in a token, so there is nothing to say.
        let model = await makePropagationModel(token: Jwt.withRoles())
        defer { cleanup(model) }
        let key = Sample.key(.entraDirectory(roleDefinitionId: Self.globalAdmin, directoryScopeId: "/administrativeUnits/au1"))
        model.active[key] = Sample.assignment(key)

        model.watchPropagation([activated(key)])
        try await settled(model, key)

        #expect(model.propagation[key] == nil)
        #expect(model.propagationNote(for: key) == nil)
    }

    @Test func deactivatingWhileItPropagatesDropsTheWatch() async throws {
        let model = await makePropagationModel(token: nil)
        defer { cleanup(model) }
        let key = Self.entraKey
        model.active[key] = Sample.assignment(key)
        model.watchPropagation([activated(key)])

        model.stopWatchingPropagation(key)

        #expect(model.propagation[key] == nil)
        #expect(model.propagationWatches[key]?.task == nil)
    }

    @Test func onlyActivatedOutcomesAreWatched() async throws {
        let model = await makePropagationModel(token: nil)
        defer { cleanup(model) }
        let key = Self.entraKey
        let pending = ActiveAssignment(roleKey: key, assignmentId: "a1", startDateTime: .now,
                                       endDateTime: nil, status: .pendingApproval)

        model.watchPropagation([ActivationOutcome(roleKey: key, result: .pendingApproval(pending))])

        #expect(model.propagation.isEmpty)
    }

    @Test func aRoleThatIsNotBeingWatchedIsNotPropagating() async throws {
        let model = await makePropagationModel(token: nil)
        defer { cleanup(model) }

        #expect(model.propagation.isEmpty)
        #expect(model.propagationNote(for: Self.entraKey) == nil)
        #expect(model.propagationTooltip(for: Self.entraKey) == nil)
        #expect(!model.isPropagating(Self.entraKey))
    }

    @Test func theUnconfirmedNoteNamesWhatToDoAboutIt() async throws {
        let model = await makePropagationModel(token: nil)
        defer { cleanup(model) }
        let key = Self.entraKey
        model.propagation[key] = .unconfirmed

        #expect(model.propagationNote(for: key) == "not in effect yet")
        #expect(model.propagationTooltip(for: key) == PropagationHints.likelyCause(.entraDirectory))
    }
}
